README.rst
Rakuten MA Python
===================

|travis| |coveralls| |pyversion| |version|  |landscape|  |license|

Rakuten MA Python (morphological analyzer) is a Python version of Rakuten MA (word segmentor + PoS Tagger) for Chinese and Japanese.

For details about Rakuten MA, See https://github.com/rakuten-nlp/rakutenma

See also http://qiita.com/yukinoi/items/925bc238185aa2fad8a7 (In Japanese)

Contributions are welcome!


Installation
==============

::

 pip install rakutenma

Example
===========

.. code:: python

 from rakutenma import RakutenMA

 # Initialize a RakutenMA instance with an empty model
 # the default ja feature set is set already
 rma = RakutenMA()

 # Let's analyze a sample sentence (from http://tatoeba.org/jpn/sentences/show/103809)
 # With a disastrous result, since the model is empty!
 print(rma.tokenize("彼は新しい仕事できっと成功するだろう。"))

 # Feed the model with ten sample sentences from tatoeba.com
 # "tatoeba.json" is available at https://github.com/rakuten-nlp/rakutenma
 import json
 tatoeba = json.load(open("tatoeba.json"))
 for i in tatoeba:
     rma.train_one(i)

 # Now what does the result look like?
 print(rma.tokenize("彼は新しい仕事できっと成功するだろう。"))

 # Initialize a RakutenMA instance with a pre-trained model
 rma = RakutenMA(phi=1024, c=0.007812)  # Specify hyperparameter for SCW (for demonstration purpose)
 rma.load("model_ja.json")

 # Set the feature hash function (15bit)
 rma.hash_func = rma.create_hash_func(15)

 # Tokenize one sample sentence
 print(rma.tokenize("うらにわにはにわにわとりがいる"));

 # Re-train the model feeding the right answer (pairs of [token, PoS tag])
 res = rma.train_one(
        [["うらにわ","N-nc"],
         ["に","P-k"],
         ["は","P-rj"],
         ["にわ","N-n"],
         ["にわとり","N-nc"],
         ["が","P-k"],
         ["いる","V-c"]])
 # The result of train_one contains:
 #   sys: the system output (using the current model)
 #   ans: answer fed by the user
 #   update: whether the model was updated
 print(res)

 # Now what does the result look like?
 print(rma.tokenize("うらにわにはにわにわとりがいる"))


NOTE
===========

Added API
--------------
As compared to original RakutenMA, following methods are added:

- RakutenMA::load(model_path)
  - Load model from JSON file

- RakutenMA::save(model_path)
  - Save model to path

misc
--------------
As initial setting, following values are set:

- rma.featset = CTYPE_JA_PATTERNS  # RakutenMA.default_featset_ja
- rma.hash_func = rma.create_hash_func(15)
- rma.tag_scheme = "SBIEO"  # if using Chinese, set "IOB2"

LICENSE
=========

Apache License version 2.0


Copyright
=============

Rakuten MA Python
(c) 2015- Yukino Ikegami. All Rights Reserved.

Rakuten MA (original)
(c) 2014 Rakuten NLP Project. All Rights Reserved.

.. |travis| image:: https://travis-ci.org/ikegami-yukino/rakutenma-python.svg?branch=master
    :target: https://travis-ci.org/ikegami-yukino/rakutenma-python
    :alt: travis-ci.org
.. |coveralls| image:: https://coveralls.io/repos/ikegami-yukino/rakutenma-python/badge.png
    :target: https://coveralls.io/r/ikegami-yukino/rakutenma-python
    :alt: coveralls.io

.. |pyversion| image:: https://img.shields.io/pypi/pyversions/rakutenma.svg

.. |version| image:: https://img.shields.io/pypi/v/rakutenma.svg
    :target: http://pypi.python.org/pypi/rakutenma/
    :alt: latest version

.. |landscape| image:: https://landscape.io/github/ikegami-yukino/rakutenma-python/master/landscape.svg?style=flat
   :target: https://landscape.io/github/ikegami-yukino/rakutenma-python/master
   :alt: Code Health

.. |license| image:: https://img.shields.io/pypi/l/rakutenma.svg
    :target: http://pypi.python.org/pypi/rakutenma/
    :alt: license


CHANGES.rst
CHANGES
=======

0.3.3 (2017-05-22)
-------------------------

- Bug fix about training

0.3.2 (2017-02-01)
-------------------------

- Use ujson when possible
- Enable POS to MeCab style
- Support Python 3.5 and 3.6

0.3 (2016-04-10)
-------------------------

- Add CUI ($ rakutenma)

0.2.2 (2016-04-09)
-------------------------

- Bundle model files (model_ja.json, model_ja_min.json)
- Support Windows

0.2 (2015-01-10)
-------------------------

- Support Python 2.6 and 2.7

0.1.1 (2015-01-08)
-------------------------

- Slightly improve performance

0.1 (2015-01-01)
-------------------------

- First release.


tests/test_scw.py
# -*- coding: utf-8 -*-
import math
from nose.tools import assert_equals, assert_almost_equals
from rakutenma import SCW, Trie


class TestTrie(object):

    def test_case1(self):
        scw = SCW(0.0, 1.0)

        assert_equals(scw.psi, 1.0)
        assert_equals(scw.zeta, 1.0)

        x = {}

        Trie.insert(x, ["a", "b", "c"], 1.0)
        Trie.insert(scw.mu, ["a", "b", "c"], 1.0)

        margin = scw.calc_margin(x, 1)
        assert_equals(margin, 1.0)

        variance = scw.calc_variance(x)
        assert_equals(variance, 1.0)

        alpha = scw.calc_alpha(margin, variance)
        assert_equals(alpha, 0.0)
        beta = scw.calc_beta(margin, variance, alpha)
        assert_equals(beta, 0.0)

        Trie.insert(x, ["a", "b", "d"], 2.0)
        Trie.insert(scw.mu, ["a", "b", "d"], 0.5)
        Trie.insert(scw.sigma, ["a", "b", "d"], 0.5)

        assert_equals(scw.calc_margin(x, -1), -2.0)
        assert_equals(scw.calc_variance(x), 3.0)

        scw.update_mu_sigma(x, 1, 1.0, 1.0)

        assert_equals(Trie.find(scw.mu, ["a", "b", "c"]), 2.0)
        assert_equals(Trie.find(scw.mu, ["a", "b", "d"]), 1.5)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "c"]), 0.0)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "d"]), -0.5)

    def test_case2(self):
        # case2: C = 1.0, phi = 2.0

        scw = SCW(2.0, 1.0)

        assert_equals(scw.psi, 3.0)
        assert_equals(scw.zeta, 5.0)

        x = {}
        Trie.insert(x, ["a", "b", "c"], 1.0)
        Trie.insert(scw.mu, ["a", "b", "c"], 1.0)

        margin = scw.calc_margin(x, 1)
        assert_equals(margin, 1.0)

        variance = scw.calc_variance(x)
        assert_equals(variance, 1.0)

        alpha = scw.calc_alpha(margin, variance)
        assert_almost_equals(alpha, (math.sqrt(24)-3)/5)

        beta = scw.calc_beta(margin, variance, alpha)
        desired = ((2 * (math.sqrt(24) - 3) / 5) /
                   (0.5 *
                    (-2 * (math.sqrt(24) - 3) / 5 +
                        math.sqrt(4 * (33 - 6 * math.sqrt(24)) / 25 + 4)) +
                    2 * (math.sqrt(24) - 3) / 5))
        assert_almost_equals(beta, desired)

        Trie.insert(x, ["a", "b", "d"], 2.0)
        scw.update_mu_sigma(x, -1, 0.2, 0.5)
        assert_equals(Trie.find(scw.mu, ["a", "b", "c"]), 0.8)
        assert_equals(Trie.find(scw.mu, ["a", "b", "d"]), -0.4)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "c"]), 0.5)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "d"]), -1.0)

    def test_prune(self):
        scw = SCW(0.0, 1.0)
        Trie.insert(scw.mu, ["a", "b", "c"], 0.5)
        Trie.insert(scw.mu, ["a", "b", "d"], 1.5)
        Trie.insert(scw.sigma, ["a", "b", "c"], 0.5)
        Trie.insert(scw.sigma, ["a", "b", "d"], 0.5)

        scw.prune(1.0, 0.8)
        assert_equals(Trie.find(scw.mu, ["a", "b", "c"]), 0)
        assert_equals(Trie.find(scw.mu, ["a", "b", "d"]), 0.5)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "c"]), 0.5)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "d"]), 0.5)

        scw.prune(1.0, 0.4)
        assert_equals(Trie.find(scw.mu, ["a", "b", "c"]), None)
        assert_equals(Trie.find(scw.mu, ["a", "b", "d"]), None)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "c"]), None)
        assert_equals(Trie.find(scw.sigma, ["a", "b", "d"]), None)


rakutenma/__init__.py
"""Rakuten MA Python

Rakuten MA Python (morphological analyzer) is a Python version of Rakuten MA
(word segmentor + PoS Tagger) for Chinese and Japanese.

Rakuten MA Python and Rakuten MA (original) are distributed under
Apache License, version 2.0. http://www.apache.org/licenses/LICENSE-2.0

Rakuten MA Python
(c) 2015- Yukino Ikegami

Rakuten MA (original)
(C) 2014 Rakuten NLP Project."""
from .rakutenma import RakutenMA, Token
from .scw import SCW
from .trie import Trie

VERSION = (0, 3, 3)
__version__ = "0.3.3"
__all__ = ["RakutenMA", "Token", "SCW", "Trie", "_DEF_LABEL", "_BEOS_LABEL",
           "FEATSET_JA", "FEATSET_ZH"]

_DEF_LABEL = "O"     # default label
_BEOS_LABEL = "_"    # label for BOS / EOS

FEATSET_JA = (
    "c0", "w0", "w1", "w9", "w2", "w8",
    "b1", "b9", "b2", "b8",
    "c1", "c9", "c2", "c8", "c3", "c7",
    "d1", "d9", "d2", "d8"
)

FEATSET_ZH = (
    "c0", "w0", "w1", "w9", "w2", "w8",
    "b1", "b9", "b2", "b8",
    "c1", "c9", "c2", "c8", "c3", "c7"
)


rakutenma/rakutenma.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import pkgutil
import os
import re
import sys
try:
    import ujson as json
    use_ujson = True
except:
    import json
    use_ujson = False
from .scw import SCW
from .trie import Trie

# character type
# from TinySegmenter http://chasen.org/~taku/software/TinySegmenter/
CTYPE_JA_PATTERNS = (
    ("[一二三四五六七八九十百千万億兆]", "S"),
    ("[一-龠々〆ヵヶ]", "C"),
    ("[ぁ-ん]", "H"),
    ("[ァ-ヴーｱ-ﾝﾞｰ]", "K"),
    ("[A-ZＡ-Ｚ]", "A"),
    ("[a-zａ-ｚ]", "a"),
    ("[0-9０-９]", "N"),
    ("[・]", "n")
)
_DEF_LABEL = "O"     # default label
_BEOS_LABEL = "_"    # label for BOS / EOS

# default feature set (first features are used as tag dictionary)
FEATSET_JA = (
    "c0", "w0", "w1", "w9", "w2", "w8",
    "b1", "b9", "b2", "b8",
    "c1", "c9", "c2", "c8", "c3", "c7",
    "d1", "d9", "d2", "d8"
)

FEATSET_ZH = (
    "c0", "w0", "w1", "w9", "w2", "w8",
    "b1", "b9", "b2", "b8",
    "c1", "c9", "c2", "c8", "c3", "c7"
)

MECAB_MAP = {'形容詞-一般': 'A-c',
'形容詞-非自立可能': 'A-dp',
'接続詞': 'C',
'代名詞': 'D',
'英単語': 'E',
'副詞': 'F',
'感動詞-一般': 'I-c',
'形状詞-一般': 'J-c',
'形状詞-タリ': 'J-tari',
'形状詞-助動詞語幹': 'J-xs',
'補助記号-AA': 'M-aa',
'補助記号-一般': 'M-c',
'補助記号-括弧閉': 'M-cp',
'補助記号-括弧開': 'M-op',
'補助記号-句点': 'M-p',
'名詞-名詞的': 'N-n',
'名詞-普通名詞': 'N-nc',
'名詞-固有名詞': 'N-pn',
'名詞-助動詞語幹': 'N-xs',
'その他': 'O',
'接頭辞': 'P',
'助詞-副助詞': 'P-fj',
'助詞-準体助詞': 'P-jj',
'助詞-格助詞': 'P-k',
'助詞-係助詞': 'P-rj',
'助詞-接続助詞': 'P-sj',
'接尾辞-形容詞的': 'Q-a',
'接尾辞-形状詞的': 'Q-j',
'接尾辞-名詞的': 'Q-n',
'接尾辞-動詞的': 'Q-v',
'連体詞': 'R',
'記号-一般': 'S-c',
'記号-文字': 'S-l',
'URL': 'U',
'動詞-一般': 'V-c',
'動詞-非自立可能': 'V-dp',
'空白': 'W',
'助動詞': 'X'}

# TDEF: transition default distribution
TDEF = Trie.insert({}, [_DEF_LABEL, _DEF_LABEL], 1.)
TDEF = Trie.insert(TDEF, [_BEOS_LABEL, _DEF_LABEL], 0.1)
TDEF = Trie.insert(TDEF, [_DEF_LABEL, _BEOS_LABEL], 0.1)

# EDEF: emission default distribution
EDEF = Trie.insert({}, [_DEF_LABEL], 0.1)
EDEF = Trie.insert(EDEF, [_BEOS_LABEL], 0.)

PY3 = sys.version_info[0] == 3


class Token():
    __slots__ = ("c", "f", "l", "t")

    def __init__(self, c="", f="", l="", t=""):
        self.c = c
        self.f = f
        self.l = l
        self.t = t

    def __repr__(self):
        return "<c=%s, f=%s, l=%s, t=%s>" % (self.c, self.f, self.l, self.t)


class RakutenMA(object):

    def __init__(self, model={}, phi=2048, c=0.003906):
        """
        Args:
            <dict> model
            <int> phi
            <float> c
        """
        self.model = model
        self.scw = SCW(phi, c)
        self.scw.mu = model.get("mu", {})
        self.scw.sigma = model.get("sigma", {})

        self.ctype_func = self.ctype_ja_default_func
        self.build_ctype_map(CTYPE_JA_PATTERNS)
        self.tag_scheme = "SBIEO"
        self.featset = FEATSET_JA
        self.hash_func = self.create_hash_func(15)

    def set_tag_scheme(self, scheme):
        self.tag_scheme = scheme

    def save(self, filename, double_precision=".6f"):
        if use_ujson:
            with open(filename, "w") as fd:
                fd.write(json.dumps(self.model, double_precision))
        else:
            json.dump(self.model, open(filename, "w"))

    def load(self, filename):
        if os.path.exists(filename):
            self.set_model(json.load(open(filename, "r")))
        elif filename == 'model_ja.json' or filename == 'model_ja.min.json':
            model_data = pkgutil.get_data('rakutenma', os.path.join('model', filename))
            if PY3:
                self.set_model(json.loads(model_data.decode('utf8')))
            else:
                self.set_model(json.loads(model_data))

    @staticmethod
    def string2hash(_str):
        """receives a string and returns a hash value for it
        from http://werxltd.com/wp/2010/05/13/javascript-implementation-of-javas-string-hashcode-method/
        Arg:
            <str> _str
        Ret:
            <int> hash
        """
        _hash = 0
        for char in _str:
            shifted = _hash << 5
            if shifted & 0x80000000:
                shifted = -(0x100000000 - shifted)
            _hash = (shifted - _hash) + ord(char)
            if _hash >= 0x80000000:
                _hash -= 0x100000000
            elif _hash <= -0x80000000:
                _hash = int("-0x" + hex(_hash)[-8:], 16)
        return _hash

    def create_hash_func(self, bits):
        """creates and returns a feature hashing function
        using the specified number of bits
        Args:
            <int> bits
        Return:
            <function> hash_func
        """
        num_feats = 0x01 << bits

        def hash_func(*x):
            _hash = self.string2hash("_".join(x))
            if _hash < 0:
                return [str(_hash % (num_feats * -1) + num_feats - 1)]
            return [str(_hash % num_feats + num_feats - 1)]
        return hash_func

    def str2csent(self, _input):
        """ convert input string to a vector of chars (csent character-sent)
        add ctypes on the way
        Args:
            <str> _input
        Return:
            <list> csent
        """
        csent = [Token(l=_BEOS_LABEL)]  # BOS
        csent += [Token(c=char, t=self.ctype_func(char)) for char in _input]
        csent.append(Token(l=_BEOS_LABEL))  # EOS
        return csent

    def add_efeats(self, csent):
        """receives csent (character-sentence) structure
        and adds the emission features to csent[i].f
        Args:
            <list> csent
        Return:
            <list> csent
        """
        csent_length = len(csent)
        _t = lambda i: csent[i] if csent_length > i >= 0 else Token()

        # feature hashing function
        if self.hash_func:
            _f = self.hash_func
        else:
            _f = lambda *x: x[0] if isinstance(x[0], list) else list(x)

        def add_ctype_feats(arr, label, ctype):
            """a helper function to add all the feature values of ctype to arr
            if ctype is a string, simply adds it to arr,
            if ctype is an array, adds all the elements to arr (used for Chinese tokenization)
            """
            if hasattr(ctype, "split"):
                arr.append(_f(label, ctype))
            else:
                arr += [_f(label, i) for i in ctype]

        for (i, x) in enumerate(csent):
            csent[i].f = []

            for feat in self.featset:
                # character type unigram
                if feat == "c0":
                    add_ctype_feats(csent[i].f, feat, _t(i).t)
                elif feat == "c1":
                    add_ctype_feats(csent[i].f, feat, _t(i+1).t)
                elif feat == "c9":
                    add_ctype_feats(csent[i].f, feat, _t(i-1).t)
                elif feat == "c2":
                    add_ctype_feats(csent[i].f, feat, _t(i+2).t)
                elif feat == "c8":
                    add_ctype_feats(csent[i].f, feat, _t(i-2).t)
                elif feat == "c3":
                    add_ctype_feats(csent[i].f, feat, _t(i+3).t)
                elif feat == "c7":
                    add_ctype_feats(csent[i].f, feat, _t(i-3).t)
                # character unigram
                elif feat == "w0":
                    csent[i].f.append(_f(feat, _t(i).c))
                elif feat == "w1":
                    csent[i].f.append(_f(feat, _t(i+1).c))
                elif feat == "w9":
                    csent[i].f.append(_f(feat, _t(i-1).c))
                elif feat == "w2":
                    csent[i].f.append(_f(feat, _t(i+2).c))
                elif feat == "w8":
                    csent[i].f.append(_f(feat, _t(i-2).c))
                elif feat == "w3":
                    csent[i].f.append(_f(feat, _t(i+3).c))
                elif feat == "w7":
                    csent[i].f.append(_f(feat, _t(i-3).c))
                # character bigram
                elif feat == "b1":
                    csent[i].f.append(_f(feat, _t(i).c, _t(i+1).c))
                elif feat == "b9":
                    csent[i].f.append(_f(feat, _t(i-1).c, _t(i).c))
                elif feat == "b2":
                    csent[i].f.append(_f(feat, _t(i+1).c, _t(i+2).c))
                elif feat == "b8":
                    csent[i].f.append(_f(feat, _t(i-2).c, _t(i-1).c))
                elif feat == "b3":
                    csent[i].f.append(_f(feat, _t(i+2).c, _t(i+3).c))
                elif feat == "b7":
                    csent[i].f.append(_f(feat, _t(i-3).c, _t(i-2).c))
                # character type bigram
                elif feat == "d1":
                    csent[i].f.append(_f(feat, _t(i).t, _t(i+1).t))
                elif feat == "d9":
                    csent[i].f.append(_f(feat, _t(i-1).t, _t(i).t))
                elif feat == "d2":
                    csent[i].f.append(_f(feat, _t(i+1).t, _t(i+2).t))
                elif feat == "d8":
                    csent[i].f.append(_f(feat, _t(i-2).t, _t(i-1).t))
                elif feat == "d3":
                    csent[i].f.append(_f(feat, _t(i+2).t, _t(i+3).t))
                elif feat == "d7":
                    csent[i].f.append(_f(feat, _t(i-3).t, _t(i-2).t))
                # character trigram
                elif feat == "t0":
                    csent[i].f.append(_f(feat, _t(i-1).c, _t(i).c, _t(i+1).c))
                elif feat == "t1":
                    csent[i].f.append(_f(feat, _t(i).c, _t(i+1).c, _t(i+2).c))
                elif feat == "t9":
                    csent[i].f.append(_f(feat, _t(i-2).c, _t(i-1).c, _t(i).c))

                else:
                    # if the feature template is a function,
                    # invoke it and add the returned value
                    if hasattr(feat, "__call__"):
                        csent[i].f.append(_f(feat(_t, i)))
                    else:
                        raise ValueError("Invalid feature specification!")
        return csent

    @staticmethod
    def calc_states0(cfeats, weights):
        """get state distribution based on emission features
        Args:
            <list> cfeat: set of feature values
            <dict> weights: feature weights (trie)
        Return:
            <dict> states0
        """
        scores0 = {}
        states0 = {}

        for (i, cfeat) in enumerate(cfeats):
            cemits = Trie.find_partial(weights, cfeat) or EDEF
            for k in cemits:
                if i == 0:
                    # tag dictionary
                    # the possible set of tags is solely defined by the first feature
                    states0[k] = True
                    scores0[k] = cemits[k]["v"]
                else:
                    scores0[k] = scores0.get(k, 0) + cemits[k]["v"]
        # replace by scores
        for s0 in states0:
            states0[s0] = scores0[s0]
        return states0

    def decode(self, csent):
        """ decode csent (character-sentence) structure based on its features
        using the Viterbi algorithm and assign lables to csent[i].l
        Args:
            <list> csent
        Return:
            <list> csent
        """

        weights = self.model.get("mu", {})
        trans = weights.get("t", TDEF.copy())

        statesp = {
            _BEOS_LABEL: {"score": 0., "path": [_BEOS_LABEL]}
        }

        states0 = {}

        for char in csent[1:]:
            states0 = self.calc_states0(char.f, weights)
            for (s0, states0_score) in states0.items():
                max_score = -float("inf")
                max_state = None
                trans0 = trans.get(s0, {})
                for sp in statesp:
                    t_score = 0.
                    if sp in trans0:
                        t_score = trans0[sp]["v"]
                    statesp_score = 0
                    if isinstance(statesp[sp], dict):
                        statesp_score = statesp[sp]["score"]
                    score = statesp_score + states0_score + t_score
                    if score > max_score:
                        max_score = score
                        max_state = sp
                if max_state and max_score > 0:
                    path = []
                    if isinstance(statesp[max_state], dict):
                        path = statesp[max_state]["path"]
                    states0[s0] = {"score": max_score, "path": path + [s0]}
            statesp = states0

        # track the path and assign to csent[i].l
        final_path = statesp[_BEOS_LABEL].get("path", {})
        for (i, x) in enumerate(csent):
            if len(final_path) > i:
                csent[i].l = final_path[i] or _DEF_LABEL
            else:
                csent[i].l = _DEF_LABEL
        return csent

    @staticmethod
    def csent2tokens(csent, scheme):
        """convert csent to tsent (mainly for final output and evaluation)
        Args:
            <list> csent
            <str> scheme: tag scheme
        Return:
            <list> tokens
        """
        tokens = []
        ctoken = None

        if scheme == "SBIEO":
            for cs in csent[1:-1]:  # Skip BOS and EOS
                head = cs.l[0]
                tail = cs.l[2:]

                if head == "B":
                    if ctoken:
                        tokens.append(ctoken)
                    ctoken = [cs.c, tail]
                elif head == "S":
                    if ctoken:
                        tokens.append(ctoken)
                    tokens.append([cs.c, tail])
                    ctoken = None
                elif head == "I":
                    ctoken = ctoken or ["", tail]
                    ctoken[0] += cs.c
                elif head == "E":
                    ctoken = ctoken or ["", tail]
                    ctoken[0] += cs.c
                    tokens.append(ctoken)
                    ctoken = None
                else:
                    if ctoken:
                        tokens.append(ctoken)
                    tokens.append([cs.c, tail])
                    ctoken = None
        elif scheme == "IOB2":
            for cs in csent[1:-1]:  # Skip BOS and EOS
                head = cs.l[0]
                tail = cs.l[2:]
                if head == "B":
                    if ctoken:
                        tokens.append(ctoken)
                    ctoken = [cs.c, tail]
                elif head == "I":
                    ctoken = ctoken or ["", tail]
                    ctoken[0] += cs.c
                else:
                    if ctoken:
                        tokens.append(ctoken)
                    tokens.append([cs.c, tail])
                    ctoken = None
        else:
            raise ValueError("Invalid tag scheme!")

        if ctoken:
            tokens.append(ctoken)
        return tokens

    def tokenize(self, _input):
        """tokenize input sentence (string)
        Args:
            <str> _input
        Return:
            <list> tokens
        """
        csent = self.str2csent(_input)
        csent = self.add_efeats(csent)
        csent = self.decode(csent)
        return self.csent2tokens(csent, self.tag_scheme)

    def tokens2csent(self, tokens, scheme):
        """convert a tsent(tokenized sentence) to the csent (character-sentence) structure
        scheme should be either SBIEO or IOB2
        Args:
            <list> tokens
            <str> scheme: tag scheme
        Return:
            <list> csent
        """
        csent = [Token(l=_BEOS_LABEL)]  # BOS

        if not isinstance(tokens[0], list):
            tokens = [tokens]

        if scheme == "SBIEO":
            for token in tokens:
                length = len(token[0])
                if length == 1:
                    csent.append(Token(c=token[0], t=self.ctype_func(token[0]),
                                       l="S-" + token[1]))
                else:
                    for j in range(length):
                        tag = "I-"
                        if j == 0:
                            tag = "B-"
                        elif j == (length - 1):
                            tag = "E-"
                        csent.append(Token(c=token[0][j],
                                           t=self.ctype_func(token[0][j]),
                                           l=tag + token[1]))
        elif scheme == "IOB2":
            for token in tokens:
                for j in range(len(token[0])):
                    tag = "B-" if j == 0 else "I-"
                    csent.append(Token(c=token[0][j:j+1],
                                       t=self.ctype_func(token[0][j]),
                                       l=tag + token[1]))
        else:
            raise ValueError("Invalid tag scheme!")
        csent.append(Token(l=_BEOS_LABEL))  # EOS
        return csent

    @staticmethod
    def csent2feats(csents):
        """receives a csent and returns a set of features
        (both transition and emission) for SCW update
        Args:
            <list> csents
        Return:
            <list> feats
        """
        feats = []
        for (i, csent) in enumerate(csents):
            feats += [f + [csent.l] for f in csent.f]
            if i > 0:
                feats.append(["t", csent.l, csents[i-1].l])
        return feats

    def train_one(self, sent, mecab=False):
        """train the current model based on a new (single) instance
        which is tsent (token-sentence)
        Args:
            <list> sent
            <bool> mecab
        Return:
            <dict> res: result
        """
        res = {}

        if mecab:
            for i in range(len(sent)):
                pos = MECAB_MAP[sent[i][1]]
                sent[i] = [sent[i][0], pos]

        # get answer feats
        ans_csent = self.tokens2csent(sent, self.tag_scheme)
        ans_csent = self.add_efeats(ans_csent)
        ans_feats = self.csent2feats(ans_csent)
        res["ans"] = self.csent2tokens(ans_csent, self.tag_scheme)

        # get system output
        sys_csent = self.decode(ans_csent)
        res["sys"] = self.csent2tokens(sys_csent, self.tag_scheme)

        # update
        if self.tokens_identical(res["ans"], res["sys"]):
            res["updated"] = False
        else:
            ans_trie = {}
            for ans_feat in ans_feats:
                ans_trie = Trie.insert(ans_trie, ans_feat, 1)
            self.scw.update(ans_trie, 1)

            sys_trie = {}
            for sys_feat in self.csent2feats(sys_csent):
                sys_trie = Trie.insert(sys_trie, sys_feat, 1)
            self.scw.update(sys_trie, -1)

            res["updated"] = True
            self.model["mu"] = self.scw.mu
            self.model["sigma"] = self.scw.sigma
        return res

    def prune(self, _lambda, sigma_th):
        """prune the model by FOBOS
        simply dispatches scw.prune
        Args:
            <float> _lambda
            <float> sigma_th
        """
        self.scw.prune(_lambda, sigma_th)
        self.model["mu"] = self.scw.mu
        self.model["sigma"] = self.scw.sigma

    def set_model(self, model):
        """set a new model
        Args:
            <dict> model
        """
        self.model = model
        self.scw.mu = model.get("mu", {})
        self.scw.sigma = model.get("sigma", {})

    @staticmethod
    def tokens2string(tokens):
        """convert tsent to a string representation
        Args:
            <list> tokens
        Return:
            <str> tokens_str
        """
        ret = ["%s [%s]" % (token[0], token[1]) for token in tokens]
        return " | ".join(ret)

    def build_ctype_map(self, pats):
        """
        Args:
            <tuple> pats
        """
        self._ctype_ja_pats = []
        for (pattern, chartype) in pats:
            self._ctype_ja_pats.append((re.compile(pattern), chartype))

    def ctype_ja_default_func(self, char):
        """default character type function for Japanese
        Args:
            <str> char
        Return:
            <str> ctype: character type
        """
        for pat in self._ctype_ja_pats:
            if pat[0].match(char):
                return pat[1]
        return "O"

    @staticmethod
    def create_ctype_chardic_func(chardic):
        """receives a chardic (object of character to set of character types
        and returns a function which uses this chardic (closure)
        mainly used for Chinese
        Args:
            <dict> chardic
        Return:
            <lambda> ctype_chardic_func
        """
        return lambda x: chardic.get(x, [])

    @staticmethod
    def tokens_identical(tokens1, tokens2):
        """checks if tokens1 and tokens2 (both tsent) are identical
        based on words and their labels
        Args:
            <list> tokens1
            <list> tokens2
        Return:
            <bool>
        """
        if len(tokens1) != len(tokens2):
            return False

        for (token1, token2) in zip(tokens1, tokens2):
            if (len(token1) < 2 or len(token2) < 2 or
               token1[0] != token2[0] or token1[1] != token2[1]):
                return False
        return True

    @staticmethod
    def tokenize_corpus(tokenize_func, corpus):
        """given a corpus (test data), tokenizes all the sentences and returns the result
        (mainly used for evaluation. see scripts/eval_ja.js and scripts/eval_zh.js)
        Args:
            <lambda> tokenize_func
            <list> corpus
        Return:
            <list> tokenized_corpus
        """
        tokenized_corpus = []
        for sent in corpus:
            sent_str = [j[0] for j in sent]
            tokenized_corpus.append(tokenize_func("".join(sent_str)))
        return tokenized_corpus

    @staticmethod
    def eval_corpus(corpus_ans, corpus_sys):
        """evaluates the corpus and computes precision, recall, and F measure
        Args:
            <list> corpus_ans
            <list> corpus_sys
        Return:
            <list> evaluation_results
        """
        tps = 0
        tokens_ans = 0
        tokens_sys = 0

        if len(corpus_ans) != len(corpus_sys):
            raise ValueError("Corpus sizes are not the same!")

        for i in range(len(corpus_ans)):
            tps += RakutenMA.count_tps(corpus_ans[i], corpus_sys[i])
            tokens_ans += len(corpus_ans[i])
            tokens_sys += len(corpus_sys[i])

        return [1.0 * tps / tokens_sys,  # precision
                1.0 * tps / tokens_ans,  # recall
                2.0 * tps / (tokens_ans + tokens_sys)]  # F1

    @staticmethod
    def count_tps(ans, _sys):
        """compare tsent (ans) and tsent (sys)
        and return the number of token-based true positives
        Args:
            <list> ans
            <list> _sys: systems's output
        Return:
            <list> evaluation_results
        """

        def token2str(token):
            if hasattr(token, "split"):
                return token
            return token[0]

        if len(ans) < len(_sys):
            min_sent = ans
            max_sent = _sys
        else:
            min_sent = _sys
            max_sent = ans

        offset = 0
        max_set = {}
        for token in max_sent:
            token_str = token2str(token)
            # attach offset in order to distinguish different tokens
            max_set[token_str + str(offset)] = True
            offset += len(token_str)

        offset = 0
        res = 0
        for token in min_sent:
            token_str = token2str(token)
            if (token_str + str(offset)) in max_set:
                res += 1
            offset += len(token_str)
        return res
